### PR TITLE
[WIP] - {bedrock}: native token bridge 

### PR DESCRIPTION
**Pull request by ethandavionlabs**
# Goals of PR

Core changes:

- add bit to the bridge 
- change  cross-chain process about ETH

### CODE REVIEW

File: packages/contracts-bedrock/contracts/libraries/Predeploys.sol [hunk 1]
```diff
@@ -6,6 +6,12 @@ pragma solidity ^0.8.0;
  * @notice Contains constant addresses for contracts that are pre-deployed to the L2 system.
  */
 library Predeploys {
+
+
+    address internal constant BVM_BIT = 0x4200000000000000000000000000000000000021;
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/libraries/Predeploys.sol:11]: Is this an address on L2? On L2, `BIT` is the native token, other than an ERC20 token.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 2]
```diff
@@ -387,6 +443,98 @@ abstract contract StandardBridge {
...
+     * @param _remoteToken Address of the corresponding token on the remote chain.
+     * @param _to          Address of the receiver.
+     * @param _amount      Amount of local tokens to deposit.
+     * @param _minGasLimit Minimum amount of gas that the bridge can be relayed with.
+     * @param _extraData   Extra data to be sent with the transaction. Note that the recipient will
+     *                     not be triggered with this data, but it will be emitted and can be used
+     *                     to identify the transaction.
+     */
+    function _initiateBridgeBITDeposit(
+        address _localToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:506]: `_localToken` is not necessary here. `BIT` token address on L1 should be a constant value.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 3]
```diff
@@ -20,6 +19,11 @@ import { MantleMintableERC20 } from "./MantleMintableERC20.sol";
 abstract contract StandardBridge {
     using SafeERC20 for IERC20;
 
+
+    uint32 constant BIT_TX = 1;
+    uint32 constant ERC20_TX = 2;
+    uint32 constant ETH_TX = 0;
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:25]: Can we move this to library? Thus in other contracts hard code types can be avoided.

Comment (ethandavionlabs) [reply packages/contracts-bedrock/contracts/universal/StandardBridge.sol:27]: ok,I will finish it

Comment (ethandavionlabs) [reply packages/contracts-bedrock/contracts/universal/StandardBridge.sol]: fixed on latest commit.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol (hunk 2 above)
Comment (ethandavionlabs) [reply packages/contracts-bedrock/contracts/universal/StandardBridge.sol:571]: finished on latest commit

File: packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol [hunk 4]
```diff
@@ -273,4 +276,49 @@ contract L2StandardBridge is StandardBridge, Semver {
         emit DepositFinalized(_remoteToken, _localToken, _from, _to, _amount, _extraData);
         super._emitERC20BridgeFinalized(_localToken, _remoteToken, _from, _to, _amount, _extraData);
     }
+
+    function _emitBITBridgeInitiated(
+        address _localToken,
+        address _remoteToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol:282]: The parameters (`_localToken` and `_remoteToken`) seem useless here.

File: packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol [hunk 5]
```diff
@@ -273,4 +276,49 @@ contract L2StandardBridge is StandardBridge, Semver {
...
+
+    /**
+     * @notice Emits the legacy DepositFinalized event followed by the ETHBridgeFinalized event.
+     *         This is necessary for backwards compatibility with the legacy bridge.
+     *
+     * @inheritdoc StandardBridge
+     */
+    function _emitBITBridgeFinalized(
+        address _localToken,
+        address _remoteToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol:307]: Ditto

File: packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol (hunk 4 above)
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/L2/L2StandardBridge.sol:282]: This function is a override function which is from StandardBridge.And L1Bridge also need to use the function

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 6]
```diff
@@ -507,6 +764,36 @@ abstract contract StandardBridge {
         emit ETHBridgeFinalized(_from, _to, _amount, _extraData);
     }
 
+    function _emitBITBridgeInitiated(
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:767]: Add comment to describe functionality and parameters.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 7]
```diff
@@ -507,6 +764,36 @@ abstract contract StandardBridge {
...
+        bytes memory _extraData
+    ) internal virtual {
+        emit BITBridgeInitiated(_local,_remote,_from, _to, _amount, _extraData);
+    }
+
+    /**
+     * @notice Emits the ETHBridgeFinalized and if necessary the appropriate legacy event when an
+     *         ETH bridge is finalized on this chain.
+     *
+     * @param _from      Address of the sender.
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:782]: missing `_localToken` and `_remoteToken`

File: packages/contracts-bedrock/contracts/L1/L1StandardBridge.sol [hunk 8]
```diff
@@ -142,6 +180,36 @@ contract L1StandardBridge is StandardBridge, Semver {
...
+        _initiateBITDeposit(
+            msg.sender,
+            msg.sender,
+            _amount,
+            _minGasLimit,
+            _extraData
+        );
+    }
+
+    function depositBITTo(
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/L1StandardBridge.sol:198]: Please reject bit deposit in `depositERC20` and `depositERC20To`

Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/L1/L1StandardBridge.sol:198]: yeah,I will add some condition judge.

File: packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol [hunk 9]
```diff
@@ -43,12 +44,17 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, Semver {
...
         uint64 _gasLimit,
         uint256 _value,
         bytes memory _data
     ) internal override {
-        PORTAL.depositTransaction{ value: _value }(_to, _value, _gasLimit, false, _data);
+        if (_type == BridgeConstants.BIT_TX){
+            PORTAL.depositTransaction{ value: 0 }(_type,_to, _value, _gasLimit, false, _data);
+        }else{
+            PORTAL.depositTransaction{ value: _value }(_type,_to, _value, _gasLimit, false, _data);
+        }
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol:57]: For eth deposit, the value will be assigned to zero here. The deposit ETH should be locked in PORTAL. If we forcefully assign zero to value, I think there will be a problem here.

File: packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol [hunk 10]
```diff
@@ -43,12 +44,17 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, Semver {
...
      */
     function _sendMessage(
+        uint32 _type,
         address _to,
         uint64 _gasLimit,
         uint256 _value,
         bytes memory _data
     ) internal override {
-        PORTAL.depositTransaction{ value: _value }(_to, _value, _gasLimit, false, _data);
+        if (_type == BridgeConstants.BIT_TX){
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol:53]: I think it is not necessary here to assign different value to `value` according to `deposit type` here.

File: packages/contracts-bedrock/contracts/L1/L1StandardBridge.sol [hunk 11]
```diff
@@ -268,7 +345,7 @@ contract L1StandardBridge is StandardBridge, Semver {
         uint32 _minGasLimit,
         bytes memory _extraData
     ) internal {
-        _initiateBridgeETH(_from, _to, msg.value, _minGasLimit, _extraData);
+        _initiateBridgeETHDeposit(Predeploys.LEGACY_ERC20_ETH,_from, _to, msg.value, _minGasLimit, _extraData);
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/L1StandardBridge.sol:348]: Is this address `LEGACY_ERC20_ETH` correct?

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol [hunk 12]
```diff
@@ -440,13 +472,13 @@ abstract contract CrossDomainMessenger is
...
-            (uint64(_message.length) * MIN_GAS_CALLDATA_OVERHEAD) +
-            // Constant overhead
-            MIN_GAS_CONSTANT_OVERHEAD;
+        // Dynamic overhead
+        ((uint64(_minGasLimit) * MIN_GAS_DYNAMIC_OVERHEAD_NUMERATOR) /
+        MIN_GAS_DYNAMIC_OVERHEAD_DENOMINATOR) +
+        // Calldata overhead
+        (uint64(_message.length) * MIN_GAS_CALLDATA_OVERHEAD) +
+        // Constant overhead
+        MIN_GAS_CONSTANT_OVERHEAD;
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:481]: If there is no real change here, please keep it unchanged.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol [hunk 13]
```diff
@@ -263,27 +265,57 @@ abstract contract CrossDomainMessenger is
...
+        emit SentMessage(_target, msg.sender, _message, messageNonce(), _minGasLimit);
+        if (_type == BridgeConstants.BIT_TX){
+            emit SentMessageExtension1(msg.sender, _amount);
+        }else{
+            emit SentMessageExtension1(msg.sender, 0);
         }
+
+    unchecked {
+        ++msgNonce;
+    }
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:318]: Add indentation

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol [hunk 14]
```diff
@@ -263,27 +265,57 @@ abstract contract CrossDomainMessenger is
...
+                    messageNonce(),
+                    msg.sender,
+                    _target,
+                    _amount,
+                    _minGasLimit,
+                    _message
+                )
+            );
+
+        } else if (_type == BridgeConstants.ETH_TX || _type == BridgeConstants.ERC20_TX) {
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:287]: How to handle `_sendMessage` if `_type` is `BIT_WITHDRAWAL_TX` or `ERC20_WITHDRAWAL_TX`

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol [hunk 15]
```diff
@@ -112,9 +112,9 @@ contract CrossDomainMessengerLegacySpacer1 {
  *         Any changes to this contract MUST result in a semver bump for contracts that inherit it.
  */
 abstract contract CrossDomainMessenger is
-    CrossDomainMessengerLegacySpacer0,
-    Initializable,
-    CrossDomainMessengerLegacySpacer1
+CrossDomainMessengerLegacySpacer0,
+Initializable,
+CrossDomainMessengerLegacySpacer1
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:117]: I suggest to keep this code unchanged

File: packages/contracts-bedrock/contracts/libraries/Predeploys.sol [hunk 16]
```diff
@@ -6,6 +6,11 @@ pragma solidity ^0.8.0;
  * @notice Contains constant addresses for contracts that are pre-deployed to the L2 system.
  */
 library Predeploys {
+
+
+    address internal constant L1_BIT = 0x6900000000000000000000000000000000000020;
+    address internal constant BVM_ETH = 0x4200000000000000000000000000000000000022;
```
[the remark this answers is not in the document]
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/libraries/Predeploys.sol:12]: sorry , I forget to consider the upgrade issue,I will find some way to keep it same.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 17]
```diff
@@ -283,15 +317,48 @@ abstract contract StandardBridge {
...
+        address _to,
+        uint256 _amount,
+        bytes calldata _extraData
+    ) public payable onlyOtherBridge {
+        require(_to != address(this), "StandardBridge: cannot send to self");
+        require(_to != address(MESSENGER), "StandardBridge: cannot send to messenger");
+
+
+
+
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:342]: Please delete multiple empty line

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 18]
```diff
@@ -20,6 +22,9 @@ import { MantleMintableERC20 } from "./MantleMintableERC20.sol";
 abstract contract StandardBridge {
     using SafeERC20 for IERC20;
 
+
+
+
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:27]: Please delete multiple empty lines

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 19]
```diff
@@ -187,9 +209,9 @@ abstract contract StandardBridge {
      *                     not be triggered with this data, but it will be emitted and can be used
      *                     to identify the transaction.
      */
-    function bridgeETH(uint32 _minGasLimit, bytes calldata _extraData) public payable onlyEOA {
-        _initiateBridgeETH(msg.sender, msg.sender, msg.value, _minGasLimit, _extraData);
-    }
+//    function bridgeETH(uint32 _minGasLimit, bytes calldata _extraData) public payable onlyEOA {
+//        _initiateBridgeETH(msg.sender, msg.sender, msg.value, _minGasLimit, _extraData);
+//    }
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:214]: If the method is useless, please just delete it.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 20]
```diff
@@ -206,14 +228,26 @@ abstract contract StandardBridge {
...
+    function bridgeL1ETHTo(
         address _to,
         uint32 _minGasLimit,
         bytes calldata _extraData
     ) public payable {
-        _initiateBridgeETH(msg.sender, _to, msg.value, _minGasLimit, _extraData);
+        _initiateBridgeETHDeposit(Predeploys.BVM_ETH,msg.sender, _to, msg.value, _minGasLimit, _extraData);
     }
 
+    function bridgeL2ETHTo(
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:239]: Please add functionality description like `bridgeL1ETHTo`

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 21]
```diff
@@ -358,7 +488,8 @@ abstract contract StandardBridge {
      *                     not be triggered with this data, but it will be emitted and can be used
      *                     to identify the transaction.
      */
-    function _initiateBridgeETH(
+    function _initiateBridgeETHDeposit(
+        address _remoteToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:492]: `L2ETH` address is a constant value. Can we leave out this parameter?

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 22]
```diff
@@ -408,6 +693,12 @@ abstract contract StandardBridge {
         uint32 _minGasLimit,
         bytes memory _extraData
     ) internal {
+
+
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:697]: Please delete the two empty lines

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 23]
```diff
@@ -372,12 +503,165 @@ abstract contract StandardBridge {
...
+     * @param _to          Address of the receiver.
+     * @param _amount      Amount of local tokens to deposit.
+     * @param _minGasLimit Minimum amount of gas that the bridge can be relayed with.
+     * @param _extraData   Extra data to be sent with the transaction. Note that the recipient will
+     *                     not be triggered with this data, but it will be emitted and can be used
+     *                     to identify the transaction.
+     */
+    function _initiateBridgeETHWithdrawal(
+        address _localToken,
+        address _remoteToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:632]: `_remoteToken` must be zero address on L1. Can we just leave out this parameter?

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 24]
```diff
@@ -372,12 +503,165 @@ abstract contract StandardBridge {
...
+            require(
+                _isCorrectTokenPair(_localToken, _remoteToken),
+                "StandardBridge: wrong remote token for Mantle Mintable ERC20 local token"
+            );
+
+            MantleMintableERC20(_localToken).burn(_from, _amount);
+        } else {
+            IERC20(_localToken).safeTransferFrom(_from, address(this), _amount);
+            deposits[_localToken][_remoteToken] = deposits[_localToken][_remoteToken] + _amount;
+        }
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:650]: If there are no `_localToken` and `_remoteToken` , the above check can be left out.

File: packages/contracts-bedrock/contracts/libraries/BridgeConstants.sol [hunk 25]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+library BridgeConstants {
+    uint32 constant ETH_TX = 0;
+    uint32 constant BIT_TX = 1;
+    uint32 constant ERC20_TX = 2;
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/libraries/BridgeConstants.sol:7]: I suggest to unify the naming style. `ETH_TX` -> `ETH_BRIDGE_TX`, `ETH_TX` -> `ETH_BRIDGE_TX`, `ERC20_TX` -> `ERC20_BRIDGE_TX`

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 26]
```diff
@@ -206,14 +228,26 @@ abstract contract StandardBridge {
      *                     not be triggered with this data, but it will be emitted and can be used
      *                     to identify the transaction.
      */
-    function bridgeETHTo(
+    function bridgeL1ETHTo(
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:231]: Why do we need to rename this method? The upper contract can just override this method.

File: packages/contracts-bedrock/contracts/L1/MantlePortal.sol [hunk 27]
```diff
@@ -246,8 +247,8 @@ contract MantlePortal is Initializable, ResourceMetering, Semver {
         // output index has been updated.
         require(
             provenWithdrawal.timestamp == 0 ||
-                L2_ORACLE.getL2Output(provenWithdrawal.l2OutputIndex).outputRoot !=
-                provenWithdrawal.outputRoot,
+            L2_ORACLE.getL2Output(provenWithdrawal.l2OutputIndex).outputRoot !=
+            provenWithdrawal.outputRoot,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/MantlePortal.sol:251]: Please revert the indent change

File: packages/contracts-bedrock/contracts/L1/MantlePortal.sol [hunk 28]
```diff
@@ -191,7 +192,7 @@ contract MantlePortal is Initializable, ResourceMetering, Semver {
      */
     // solhint-disable-next-line ordering
     receive() external payable {
-        depositTransaction(msg.sender, msg.value, RECEIVE_DEFAULT_GAS_LIMIT, false, bytes(""));
+        depositTransaction(msg.sender, 0, RECEIVE_DEFAULT_GAS_LIMIT, false, bytes(""));
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/MantlePortal.sol:195]: ETH on L2 is an ERC20 token. So it is right the replace `msg.value` with 0. However, I think `bytes("")` is wrong here.

File: packages/contracts-bedrock/contracts/L1/MantlePortal.sol [hunk 29]
```diff
@@ -294,8 +295,8 @@ contract MantlePortal is Initializable, ResourceMetering, Semver {
      * @param _tx Withdrawal transaction to finalize.
      */
     function finalizeWithdrawalTransaction(Types.WithdrawalTransaction memory _tx)
-        external
-        whenNotPaused
+    external
+    whenNotPaused
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/MantlePortal.sol:299]: Please revert the indent change

File: packages/contracts-bedrock/contracts/L2/L2ERC721Bridge.sol [hunk 30]
```diff
@@ -118,7 +118,7 @@ contract L2ERC721Bridge is ERC721Bridge, Semver {
 
         // Send message to L1 bridge
         // slither-disable-next-line reentrancy-events
-        MESSENGER.sendMessage(OTHER_BRIDGE, message, _minGasLimit);
+        MESSENGER.sendMessage(0,0,OTHER_BRIDGE, message, _minGasLimit);
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L2/L2ERC721Bridge.sol:121]: Please specify a type for ERC721 bridge.

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol [hunk 31]
```diff
@@ -283,15 +317,48 @@ abstract contract StandardBridge {
...
      * @param _to        Address of the receiver.
      * @param _amount    Amount of ETH being bridged.
      * @param _extraData Extra data to be sent with the transaction. Note that the recipient will
      *                   not be triggered with this data, but it will be emitted and can be used
      *                   to identify the transaction.
      */
-    function finalizeBridgeETH(
+    function finalizeBridgeETHDeposit(
+        address _localToken,
+        address _remoteToken,
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/universal/StandardBridge.sol:331]: According to my understand, `_localToken` must be BVM_ETH address and `_remoteToken` must be zero address. So why do we need the two parameter?

File: packages/contracts-bedrock/contracts/libraries/BridgeConstants.sol (hunk 25 above)
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/libraries/BridgeConstants.sol:7]: fixed

File: packages/contracts-bedrock/contracts/libraries/Predeploys.sol [hunk 32]
```diff
@@ -6,6 +6,11 @@ pragma solidity ^0.8.0;
  * @notice Contains constant addresses for contracts that are pre-deployed to the L2 system.
  */
 library Predeploys {
+
+
+    address internal constant L1_BIT = 0x6400000000000000000000000000000000000022;
```
[the remark this answers is not in the document]
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/libraries/Predeploys.sol]: fixed now

File: packages/contracts-bedrock/contracts/universal/StandardBridge.sol (hunk 6 above)
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/universal/StandardBridge.sol]: fixed

File: packages/contracts-bedrock/contracts/L2/BVM_ETH.sol [hunk 33]
```diff
@@ -0,0 +1,23 @@
...
+
+/* Contract Imports */
+import { MantleMintableERC20 } from "../universal/MantleMintableERC20.sol";
+
+/**
+ * @title BVM_ETH
+ * @dev The ETH predeploy provides an ERC20 interface for ETH deposited to Layer 2. Note that
+ * unlike on Layer 1, Layer 2 accounts do not have a balance field.
+ */
+contract BVM_ETH is MantleMintableERC20 {
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L2/BVM_ETH.sol:15]: Is this contract just for test? If so please move it under test directory.

File: packages/contracts-bedrock/contracts/L2/L2CrossDomainMessenger.sol [hunk 34]
```diff
@@ -49,14 +50,26 @@ contract L2CrossDomainMessenger is CrossDomainMessenger, Semver {
...
+        }else if (_type == BridgeConstants.MNT_WITHDRAWAL_TX){
+            L2ToL1MessagePasser(payable(Predeploys.L2_TO_L1_MESSAGE_PASSER)).initiateWithdrawal{
+                    value: _value
+                }(_type,_value,_to, _gasLimit, _data);
+        }else{
+            L2ToL1MessagePasser(payable(Predeploys.L2_TO_L1_MESSAGE_PASSER)).initiateWithdrawal{
+                    value: _value
+                }(_type,_value,_to, _gasLimit, _data);
+        }
+
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L2/L2CrossDomainMessenger.sol:72]: The external ETH withdraw is not payable. So the value must be zero. For other case, you just specify the same code. So I think there is no necessary to just `_type` here.

File: packages/contracts-bedrock/contracts/L2/L2ERC721Bridge.sol (hunk 30 above)
Comment (HaoyangLiu) [reply packages/contracts-bedrock/contracts/L2/L2ERC721Bridge.sol:121]: This comment is still not resolved

File: packages/contracts-bedrock/contracts/L1/MantlePortal.sol (hunk 28 above)
Comment (HaoyangLiu) [reply packages/contracts-bedrock/contracts/L1/MantlePortal.sol:195]: This comment is still not resolved. Does you test cases cover this case?

File: packages/contracts-bedrock/contracts/L1/MantlePortal.sol (hunk 27 above)
Comment (HaoyangLiu) [reply packages/contracts-bedrock/contracts/L1/MantlePortal.sol:251]: The indent change is still there

File: packages/contracts-bedrock/contracts/L1/TestBitToken.sol [hunk 35]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract BitTokenERC20 is ERC20 {
```
Comment (HaoyangLiu) [inline packages/contracts-bedrock/contracts/L1/TestBitToken.sol:6]: Please move the test token under test directory.

File: packages/contracts-bedrock/contracts/L2/BVM_ETH.sol (hunk 33 above)
Comment (Ethanncnm) [reply packages/contracts-bedrock/contracts/L2/BVM_ETH.sol:15]: it is used for dev-network up

**Pull request closed without merging**